rclrs/src/node/mod.rs

```diff
@@ -0,0 +1,177 @@
+use crate::error::{RclError, RclResult, ToRclResult};
+use crate::qos::QoSProfile;
+use crate::{Context, ContextHandle, Handle};
+use crate::rcl_bindings::*;
+use std::cell::{Ref, RefCell, RefMut};
+use std::ffi::CString;
+use std::rc::{Rc, Weak};
+
+pub mod publisher;
+pub use self::publisher::*;
+pub mod subscription;
+pub use self::subscription::*;
+
+pub struct NodeHandle(RefCell<rcl_node_t>);
+
+impl<'a> Handle<rcl_node_t> for &'a NodeHandle {
+    type DerefT = Ref<'a, rcl_node_t>;
+    type DerefMutT = RefMut<'a, rcl_node_t>;
+
+    fn get(self) -> Self::DerefT {
+        self.0.borrow()
+    }
+
+    fn get_mut(self) -> Self::DerefMutT {
+        self.0.borrow_mut()
+    }
+}
+
+impl Drop for NodeHandle {
+    fn drop(&mut self) {
+        let handle = &mut *self.get_mut();
+        unsafe {
+            rcl_node_fini(handle as *mut _).unwrap();
+        }
+    }
+}
+
+pub struct Node {
+    handle: Rc<NodeHandle>,
+    context: Rc<ContextHandle>,
+    subscriptions: Vec<Weak<SubscriptionBase>>,
+}
+
+impl Node {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(node_name: &str, context: &Context) -> RclResult<Node> {
+        Self::new_with_namespace(node_name, "", context)
+    }
+
+    pub fn new_with_namespace(
+        node_name: &str,
+        node_ns: &str,
+        context: &Context,
+    ) -> RclResult<Node> {
+        let raw_node_name = CString::new(node_name).unwrap();
+        let raw_node_ns = CString::new(node_ns).unwrap();
+
+        let mut node_handle = unsafe { rcl_get_zero_initialized_node() };
+        let context_handle = &mut *context.handle.get_mut();
+
+        unsafe {
+            let node_options = rcl_node_get_default_options();
+            rcl_node_init(
+                &mut node_handle as *mut _,
+                raw_node_name.as_ptr(),
+                raw_node_ns.as_ptr(),
+                context_handle as *mut _,
+                &node_options as *const _,
+            )
+            .ok()?;
+        }
+
+        let handle = Rc::new(NodeHandle(RefCell::new(node_handle)));
+
+        Ok(Node {
+            handle,
+            context: context.handle.clone(),
+            subscriptions: vec![],
+        })
+    }
+
+    pub fn create_publisher<T>(&self, topic: &str, qos: QoSProfile) -> RclResult<Publisher<T>>
+    where
+        T: rclrs_common::traits::MessageDefinition<T>,
+    {
+        Publisher::<T>::new(self, topic, qos)
+    }
+
+    pub fn create_subscription<T>(
+        &mut self,
+        topic: &str,
+        qos: QoSProfile,
+        callback: fn(&T),
+    ) -> RclResult<Rc<Subscription<T>>>
+    where
+        T: rclrs_common::traits::MessageDefinition<T> + Default,
+    {
+        let subscription = Rc::new(Subscription::<T>::new(self, topic, qos, callback)?);
```

Comment:
I thought it was closer to the rclcpp implementation as I believe publishers and subscribers hold a shared_ptr to the node. Do you suggest to have node's lifetime depending on the lifetime of the context too ?

Reply:
That's because C++ doesn't have lifetimes ;-)

Don't worry, just add a TODO comment stating that this will eventually be replaced by lifetimes and we'll address that in a subsequent pull request.

Reply:
Resolved in 5589047